File: apps/desktop/src/lib/tauri/bitcoin.ts
/**
 * Tauri IPC wrappers for Bitcoin (BDK) commands
 *
 * These functions wrap the Tauri invoke API for type-safe communication
 * with the Rust backend Bitcoin module.
 */

import { invoke } from "@tauri-apps/api/core";

// =============================================================================
// Types matching Rust backend
// =============================================================================

export type BitcoinNetwork = "mainnet" | "testnet" | "signet" | "regtest";

export type TransactionDirection = "received" | "sent" | "internal";

export type SyncStage = "connecting" | "scanning" | "updating" | "complete" | "failed";

export interface BitcoinBalance {
  /** Confirmed balance in satoshis */
  confirmed: number;
  /** Unconfirmed balance (pending transactions) in satoshis */
  unconfirmed: number;
  /** Immature balance (coinbase rewards not yet spendable) in satoshis */
  immature: number;
  /** Total trusted balance (confirmed + trusted pending) in satoshis */
  trusted_spendable: number;
}

export interface ConfirmationStatus {
  type: "confirmed" | "unconfirmed";
  block_height?: number;
  block_time?: number;
  confirmations?: number;
}

export interface BitcoinTransaction {
  /** Transaction ID (txid) */
  txid: string;
  /** Transaction direction */
  direction: TransactionDirection;
  /** Amount transferred in satoshis (positive for received, negative for sent) */
  amount_sats: number;
  /** Transaction fee in satoshis (only for sent transactions) */
  fee_sats: number | null;
  /** Confirmation status */
  status: ConfirmationStatus;
  /** Transaction timestamp (block time if confirmed, first seen if unconfirmed) */
  timestamp: number | null;
  /** Addresses involved */
  addresses: string[];
  /** Raw transaction size in bytes */
  size: number | null;
  /** Virtual size in vbytes (for fee calculation) */
  vsize: number | null;
}

export interface UtxoInfo {
  /** Transaction ID containing this UTXO */
  txid: string;
  /** Output index within the transaction */
  vout: number;
  /** Amount in satoshis */
  amount_sats: number;
  /** Address that owns this UTXO */
  address: string;
  /** Whether this UTXO is confirmed */
  is_confirmed: boolean;
  /** Block height if confirmed */
  block_height: number | null;
}

export interface SyncProgress {
  /** Current sync stage */
  stage: SyncStage;
  /** Progress percentage (0.0 - 100.0) */
  progress: number;
  /** Optional message describing current operation */
  message: string | null;
}

export interface FeeEstimate {
  /** Fee rate in satoshis per vbyte */
  sat_per_vbyte: number;
  /** Target confirmation blocks */
  target_blocks: number;
}

export interface SendTransactionResult {
  /** Transaction ID */
  txid: string;
  /** Raw signed transaction hex */
  tx_hex: string;
  /** Fee paid in satoshis */
  fee_sats: number | null;
  /** Virtual size in vbytes */
  vsize: number;
  /** Whether the transaction was broadcast */
  broadcast: boolean;
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Convert satoshis to BTC
 */
export function satsToBtc(sats: number): number {
  return sats / 100_000_000;
}

/**
 * Convert BTC to satoshis
 */
export function btcToSats(btc: number): number {
  return Math.round(btc * 100_000_000);
}

/**
 * Format satoshis as BTC string with specified decimals
 */
export function formatBtc(sats: number, decimals: number = 8): string {
  return satsToBtc(sats).toFixed(decimals);
}

// =============================================================================
// Wallet Management
// =============================================================================

/**
 * Create a Bitcoin wallet from seed
 *
 * @param walletId - Unique identifier for the wallet
 * @param seed - 64-byte seed (from mnemonic)
 * @param account - BIP84 account index (default: 0)
 */
export async function createBitcoinWallet(
  walletId: string,
  seed: number[],
  account: number = 0
): Promise<string> {
  return invoke<string>("bitcoin_create_wallet", {
    walletId,
    seed,
    account,
  });
}

/**
 * Create a watch-only Bitcoin wallet from xpub
 *
 * @param walletId - Unique identifier for the wallet
 * @param xpub - Extended public key (xpub, zpub, etc.)
 */
export async function createBitcoinWatchWallet(
  walletId: string,
  xpub: string
): Promise<string> {
  return invoke<string>("bitcoin_create_watch_wallet", {
    walletId,
    xpub,
  });
}

/**
 * Check if a Bitcoin wallet exists
 */
export async function bitcoinWalletExists(walletId: string): Promise<boolean> {
  return invoke<boolean>("bitcoin_wallet_exists", { walletId });
}

/**
 * Initialize a Bitcoin wallet from the cached seed in WalletManager
 *
 * This should be called after creating or importing an HD wallet
 * that includes Bitcoin, to initialize the BDK wallet for syncing.
 *
 * @param walletId - Wallet ID from createHDWallet/importHDWallet
 * @param account - BIP84 account index (default: 0)
 */
export async function initBitcoinFromCachedSeed(
  walletId: string,
  account: number = 0
): Promise<string> {
  return invoke<string>("bitcoin_init_from_cached_seed", {
    walletId,
    account,
  });
}

// =============================================================================
// Synchronization
// =============================================================================

/**
 * Sync a Bitcoin wallet with the blockchain
 *
 * This will connect to Electrum servers and fetch the latest
 * transaction data for the wallet.
 */
export async function syncBitcoinWallet(walletId: string): Promise<SyncProgress> {
  return invoke<SyncProgress>("bitcoin_sync_wallet", { walletId });
}

// =============================================================================
// Balance & Transactions
// =============================================================================

/**
 * Get Bitcoin wallet balance
 */
export async function getBitcoinBalance(walletId: string): Promise<BitcoinBalance> {
  return invoke<BitcoinBalance>("bitcoin_get_balance", { walletId });
}

/**
 * Get Bitcoin transaction history
 */
export async function getBitcoinTransactions(
  walletId: string
): Promise<BitcoinTransaction[]> {
  return invoke<BitcoinTransaction[]>("bitcoin_get_transactions", { walletId });
}

/**
 * Get Bitcoin UTXOs
 */
export async function getBitcoinUtxos(walletId: string): Promise<UtxoInfo[]> {
  return invoke<UtxoInfo[]>("bitcoin_get_utxos", { walletId });
}

// =============================================================================
// Addresses
// =============================================================================

/**
 * Get a new Bitcoin receiving address
 */
export async function getBitcoinNewAddress(walletId: string): Promise<string> {
  return invoke<string>("bitcoin_get_new_address", { walletId });
}

// =============================================================================
// Fee Estimation
// =============================================================================

/**
 * Estimate fee for Bitcoin transaction
 *
 * @param targetBlocks - Target number of blocks for confirmation (default: 6)
 */
export async function estimateBitcoinFee(
  targetBlocks: number = 6
): Promise<FeeEstimate> {
  return invoke<FeeEstimate>("bitcoin_estimate_fee", { targetBlocks });
}

// =============================================================================
// Network Info
// =============================================================================

/**
 * Get current Bitcoin network
 */
export async function getBitcoinNetwork(): Promise<BitcoinNetwork> {
  return invoke<BitcoinNetwork>("bitcoin_get_network");
}

// =============================================================================
// Single Address Queries (for watch-only addresses)
// =============================================================================

/**
 * Get balance for a single Bitcoin address
 *
 * Use this for watch-only single addresses. Queries Electrum directly
 * without needing a BDK wallet.
 *
 * @param address - Bitcoin address (bc1..., 1..., 3...)
 */
export async function getBitcoinAddressBalance(
  address: string
): Promise<BitcoinBalance> {
  return invoke<BitcoinBalance>("bitcoin_get_address_balance", { address });
}

/**
 * Get transaction history for a single Bitcoin address
 *
 * Use this for watch-only single addresses. Queries Electrum directly
 * without needing a BDK wallet.
 *
 * @param address - Bitcoin address (bc1..., 1..., 3...)
 */
export async function getBitcoinAddressTransactions(
  address: string
): Promise<BitcoinTransaction[]> {
  return invoke<BitcoinTransaction[]>("bitcoin_get_address_transactions", {
    address,
  });
}

/**
 * Check if a string is a Bitcoin address (vs an xpub)
 */
export function isBitcoinAddress(input: string): boolean {
  return (
    input.startsWith("bc1") ||
    input.startsWith("tb1") ||
    input.startsWith("1") ||
    input.startsWith("3") ||
    input.startsWith("m") ||
    input.startsWith("n") ||
    input.startsWith("2")
  );
}

// =============================================================================
// Send Transaction
// =============================================================================

/**
 * Send Bitcoin to a recipient address
 *
 * Creates, signs, and broadcasts a Bitcoin transaction.
 * Only works with HD wallets that have signing capability.
 *
 * @param walletId - Wallet ID
 * @param recipientAddress - Bitcoin address to send to
 * @param amountSats - Amount to send in satoshis
 * @param feeRate - Fee rate in sat/vB
 * @param broadcast - Whether to broadcast the transaction (default: true)
 */
export async function sendBitcoinTransaction(
  walletId: string,
  recipientAddress: string,
  amountSats: number,
  feeRate: number,
  broadcast: boolean = true
): Promise<SendTransactionResult> {
  return invoke<SendTransactionResult>("bitcoin_send_transaction", {
    walletId,
    recipientAddress,
    amountSats,
    feeRate,
    broadcast,
  });
}

/**
 * Validate a Bitcoin address for the current network
 *
 * @param address - Bitcoin address to validate
 * @returns true if valid for current network
 */
export async function validateBitcoinAddress(address: string): Promise<boolean> {
  return invoke<boolean>("bitcoin_validate_address", { address });
}

File: apps/desktop/src-tauri/src/commands/bitcoin.rs
//! Bitcoin-specific Tauri commands
//!
//! Exposes BDK wallet functionality to the frontend.

use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use std::sync::Arc;
use tauri::State;
use tracing::{debug, error, info};

use crate::error::{Error, Result};
use crate::wallet::bitcoin::{
    BitcoinAdapter, BitcoinBalance, BitcoinConfig, BitcoinNetwork, BitcoinTransaction, FeeEstimate,
    SendTransactionResult, SyncProgress, UtxoInfo,
};
use super::wallet::get_wallet_manager;

/// Bitcoin adapter state for Tauri
pub struct BitcoinState {
    adapter: Arc<BitcoinAdapter>,
}

impl BitcoinState {
    /// Create a new Bitcoin state with mainnet configuration
    pub fn new(data_dir: PathBuf) -> Self {
        let bitcoin_db_dir = data_dir.join("bitcoin");
        std::fs::create_dir_all(&bitcoin_db_dir).ok();

        Self {
            adapter: Arc::new(BitcoinAdapter::mainnet(bitcoin_db_dir)),
        }
    }

    /// Create with custom configuration
    pub fn with_config(config: BitcoinConfig, data_dir: PathBuf) -> Self {
        let bitcoin_db_dir = data_dir.join("bitcoin");
        std::fs::create_dir_all(&bitcoin_db_dir).ok();

        Self {
            adapter: Arc::new(BitcoinAdapter::new(config, bitcoin_db_dir)),
        }
    }

    /// Get reference to the adapter
    pub fn adapter(&self) -> &BitcoinAdapter {
        &self.adapter
    }
}

// ============================================================================
// Request/Response Types
// ============================================================================

#[derive(Debug, Deserialize)]
pub struct CreateBitcoinWalletRequest {
    pub wallet_id: String,
    pub seed: Vec<u8>,
    pub account: Option<u32>,
}

#[derive(Debug, Deserialize)]
pub struct CreateWatchWalletRequest {
    pub wallet_id: String,
    pub xpub: String,
}

#[derive(Debug, Serialize)]
pub struct BitcoinWalletInfo {
    pub wallet_id: String,
    pub network: BitcoinNetwork,
    pub balance: BitcoinBalance,
}

// ============================================================================
// Tauri Commands
// ============================================================================

/// Create a Bitcoin wallet from seed
#[tauri::command]
pub async fn bitcoin_create_wallet(
    state: State<'_, BitcoinState>,
    wallet_id: String,
    seed: Vec<u8>,
    account: Option<u32>,
) -> Result<String> {
    info!("Creating Bitcoin wallet: {}", wallet_id);

    if seed.len() != 64 {
        return Err(Error::Bitcoin(format!(
            "Invalid seed length: expected 64 bytes, got {}",
            seed.len()
        )));
    }

    let seed_array: [u8; 64] = seed.try_into().map_err(|_| {
        Error::Bitcoin("Failed to convert seed to fixed array".to_string())
    })?;

    state
        .adapter()
        .create_wallet_from_seed(&seed_array, &wallet_id, account.unwrap_or(0))?;

    Ok(wallet_id)
}

/// Create a watch-only Bitcoin wallet from xpub
#[tauri::command]
pub async fn bitcoin_create_watch_wallet(
    state: State<'_, BitcoinState>,
    wallet_id: String,
    xpub: String,
) -> Result<String> {
    info!("Creating watch-only Bitcoin wallet: {}", wallet_id);

    state.adapter().create_watch_wallet(&xpub, &wallet_id)?;

    Ok(wallet_id)
}

/// Sync a Bitcoin wallet with the blockchain
#[tauri::command]
pub async fn bitcoin_sync_wallet(
    state: State<'_, BitcoinState>,
    wallet_id: String,
) -> Result<SyncProgress> {
    info!("Syncing Bitcoin wallet: {}", wallet_id);

    let mut wallet = state.adapter().load_wallet(&wallet_id)?;
    let progress = state.adapter().sync_wallet(&mut wallet)?;

    Ok(progress)
}

/// Get Bitcoin wallet balance
#[tauri::command]
pub async fn bitcoin_get_balance(
    state: State<'_, BitcoinState>,
    wallet_id: String,
) -> Result<BitcoinBalance> {
    debug!("Getting balance for wallet: {}", wallet_id);

    let wallet = state.adapter().load_wallet(&wallet_id)?;
    let balance = state.adapter().get_balance(&wallet)?;

    Ok(balance)
}

/// Get Bitcoin transaction history
#[tauri::command]
pub async fn bitcoin_get_transactions(
    state: State<'_, BitcoinState>,
    wallet_id: String,
) -> Result<Vec<BitcoinTransaction>> {
    debug!("Getting transactions for wallet: {}", wallet_id);

    let wallet = state.adapter().load_wallet(&wallet_id)?;
    let transactions = state.adapter().get_transactions(&wallet)?;

    Ok(transactions)
}

/// Get Bitcoin UTXOs
#[tauri::command]
pub async fn bitcoin_get_utxos(
    state: State<'_, BitcoinState>,
    wallet_id: String,
) -> Result<Vec<UtxoInfo>> {
    debug!("Getting UTXOs for wallet: {}", wallet_id);

    let wallet = state.adapter().load_wallet(&wallet_id)?;
    let utxos = state.adapter().get_utxos(&wallet)?;

    Ok(utxos)
}

/// Estimate fee for Bitcoin transaction
#[tauri::command]
pub async fn bitcoin_estimate_fee(
    state: State<'_, BitcoinState>,
    target_blocks: Option<u32>,
) -> Result<FeeEstimate> {
    debug!("Estimating fee for {} blocks", target_blocks.unwrap_or(6));

    let estimate = state
        .adapter()
        .estimate_fee(target_blocks.unwrap_or(6))?;

    Ok(estimate)
}

/// Get a new Bitcoin receiving address
#[tauri::command]
pub async fn bitcoin_get_new_address(
    state: State<'_, BitcoinState>,
    wallet_id: String,
) -> Result<String> {
    debug!("Getting new address for wallet: {}", wallet_id);

    let mut wallet = state.adapter().load_wallet(&wallet_id)?;
    let address = state.adapter().get_new_address(&mut wallet)?;

    Ok(address)
}

/// Get current Bitcoin network
#[tauri::command]
pub async fn bitcoin_get_network(state: State<'_, BitcoinState>) -> Result<BitcoinNetwork> {
    Ok(state.adapter().network())
}

/// Check if a Bitcoin wallet exists
#[tauri::command]
pub async fn bitcoin_wallet_exists(
    state: State<'_, BitcoinState>,
    wallet_id: String,
) -> Result<bool> {
    match state.adapter().load_wallet(&wallet_id) {
        Ok(_) => Ok(true),
        Err(Error::Bitcoin(msg)) if msg.contains("not found") => Ok(false),
        Err(e) => Err(e),
    }
}

/// Initialize a Bitcoin wallet from the cached seed in WalletManager
///
/// This should be called after create_hd_wallet or import_hd_wallet
/// to initialize the BDK wallet for Bitcoin syncing.
#[tauri::command]
pub async fn bitcoin_init_from_cached_seed(
    state: State<'_, BitcoinState>,
    wallet_id: String,
    account: Option<u32>,
) -> Result<String> {
    info!("Initializing Bitcoin wallet from cached seed: {}", wallet_id);

    // Get the cached seed from WalletManager
    let wallet_manager = get_wallet_manager();
    let seed = wallet_manager
        .storage()
        .get_seed(&wallet_id)
        .map_err(|e| Error::Bitcoin(format!("Failed to get cached seed: {}", e)))?;

    // Create the BDK wallet
    state
        .adapter()
        .create_wallet_from_seed(&seed, &wallet_id, account.unwrap_or(0))?;

    info!("Bitcoin wallet initialized: {}", wallet_id);
    Ok(wallet_id)
}

/// Get balance for a single Bitcoin address (direct Electrum query)
///
/// Use this for watch-only single addresses instead of bitcoin_get_balance
#[tauri::command]
pub async fn bitcoin_get_address_balance(
    state: State<'_, BitcoinState>,
    address: String,
) -> Result<BitcoinBalance> {
    info!("Getting balance for address: {}", address);

    let balance = state.adapter().get_address_balance(&address)?;

    Ok(balance)
}

/// Get transaction history for a single Bitcoin address (direct Electrum query)
///
/// Use this for watch-only single addresses instead of bitcoin_get_transactions
#[tauri::command]
pub async fn bitcoin_get_address_transactions(
    state: State<'_, BitcoinState>,
    address: String,
) -> Result<Vec<BitcoinTransaction>> {
    info!("Getting transactions for address: {}", address);

    let transactions = state.adapter().get_address_transactions(&address)?;

    Ok(transactions)
}

/// Send Bitcoin to a recipient address
///
/// Creates, signs, and broadcasts a Bitcoin transaction.
/// Only works with HD wallets that have signing capability.
#[tauri::command]
pub async fn bitcoin_send_transaction(
    state: State<'_, BitcoinState>,
    wallet_id: String,
    recipient_address: String,
    amount_sats: u64,
    fee_rate: f32,
    broadcast: Option<bool>,
) -> Result<SendTransactionResult> {
    info!(
        "Sending {} sats from {} to {} at {} sat/vB",
        amount_sats, wallet_id, recipient_address, fee_rate
    );

    let mut wallet = state.adapter().load_wallet(&wallet_id)?;

    let result = state.adapter().create_and_send_transaction(
        &mut wallet,
        &recipient_address,
        amount_sats,
        fee_rate,
        broadcast.unwrap_or(true),
    )?;

    Ok(result)
}

/// Validate a Bitcoin address for the current network
#[tauri::command]
pub async fn bitcoin_validate_address(
    state: State<'_, BitcoinState>,
    address: String,
) -> Result<bool> {
    use bitcoin::Address;

    // Parse address
    let parsed = address.parse::<Address<bitcoin::address::NetworkUnchecked>>();

    if parsed.is_err() {
        return Ok(false);
    }

    // Check network
    let network: bitcoin::Network = state.adapter().network().into();
    let is_valid = parsed.unwrap().is_valid_for_network(network);

    Ok(is_valid)
}
